File: gui/gridmenu.py
__author__ = "Olof Karlsson"
__license__ = "All Rights Reserved"

import pygame
import gui.menu as menu
import settings.settings as settings

"""

This is a subclass of Menu that positions it's items in a grid (table) like fashion, instead of a vertical list like in Menu.
The user can specify the max number of columns, and the GridMenu takes care of positioning each of its items accordingly.

Other than that, it's a subclass of menu so it works pretty much like menu.

"""

class GridMenu(menu.Menu):

	offset = 2 * settings.GAME_SCALE

	def __init__(self, max_number_of_columns = 3):
		# We call the superconstructor, ofcourse.
		super(GridMenu, self).__init__()

		# This is the max number of items that are displayed next to each other before a new row is added.
		self.max_number_of_columns = max_number_of_columns

		# This is the offset between each item in the menu.
		self.offset = GridMenu.offset

		# We setup a few default variables.
		self.current_row_size = 0
		self.current_row_position = self.y

	def position_item(self, item):
		self.current_row_size += 1

		# Populate the grid with the item, and store the new current row size and current row position.
		size_and_position = self.populate_grid(item, self.current_row_size, self.current_row_position)
		self.current_row_size = size_and_position[0]
		self.current_row_position = size_and_position[1]

	def populate_grid(self, item, row_size, row_position):
		# Populates the grid according to the max_number_of_columns.
		if row_size > self.max_number_of_columns:
			# If the row size is larger than the max number of columns, we position the item below the last row and at the beginning of the new row.
			row_size = 1
			row_position = row_position + item.get_height() + self.offset

		if row_size > 1:
			# If the row size is larger than one, we position the item to the right of the last item in the last row.
			item.x = self.x + (row_size - 1) * item.get_width() + (row_size - 1) * self.offset
			item.y = row_position
		else:
			# Otherwise, we position the item to the beginning of the row.
			item.x = self.x
			item.y = row_position

		# We return the changed row_size and row_position.
		return (row_size, row_position)
		
	def cleanup(self):
		# The superclass takes care of repositioning all items, we just set a few default values.
		self.current_row_size = 0
		self.current_row_position = self.y

		super(GridMenu, self).cleanup()
				

File: c-port/src/ui/gridmenu.c
/**
 * @file gridmenu.c
 * @brief Implementation of the GridMenu UI component.
 */

#include "ui/gridmenu.h"
#include "ui/item.h"

// Default offset between items
#define GRID_MENU_DEFAULT_OFFSET 2.0f

GridMenu InitGridMenu(float x, float y, int position, int max_number_of_columns) {
    GridMenu menu;
    
    // Initialize base menu
    menu.base = InitMenu(x, y, position);
    
    // Initialize grid properties
    menu.max_number_of_columns = max_number_of_columns;
    menu.offset = GRID_MENU_DEFAULT_OFFSET;
    
    // Initialize current state
    menu.current_row_size = 0;
    menu.current_row_position = y;
    
    return menu;
}

void UnloadGridMenu(GridMenu* menu) {
    // Unload base menu
    UnloadMenu(&menu->base);
}

void PopulateGridMenu(GridMenu* menu, Item* item, int row_size, float row_position, int* new_row_size, float* new_row_position) {
    // Populates the grid according to the max_number_of_columns
    if (row_size > menu->max_number_of_columns) {
        // If the row size is larger than the max number of columns,
        // we position the item below the last row and at the beginning of the new row
        row_size = 1;
        row_position = row_position + GetItemHeight(item) + menu->offset;
    }
    
    if (row_size > 1) {
        // If the row size is larger than one, we position the item to the right of the last item in the last row
        item->x = menu->base.x + (row_size - 1) * GetItemWidth(item) + (row_size - 1) * menu->offset;
        item->y = row_position;
    } else {
        // Otherwise, we position the item to the beginning of the row
        item->x = menu->base.x;
        item->y = row_position;
    }
    
    // Return the changed row_size and row_position
    *new_row_size = row_size;
    *new_row_position = row_position;
}

void PositionGridMenuItem(GridMenu* menu, Item* item) {
    // Increment the current row size
    menu->current_row_size++;
    
    // Populate the grid with the item, and update the current row size and position
    PopulateGridMenu(menu, item, menu->current_row_size, menu->current_row_position, 
                    &menu->current_row_size, &menu->current_row_position);
}

void AddGridMenuItem(GridMenu* menu, Item* item, MenuItemFunction function, void* data) {
    // Add the item to the base menu
    AddMenuItem(&menu->base, item, function, data);
    
    // Position the item according to grid menu layout
    PositionGridMenuItem(menu, item);
}

void CleanupGridMenu(GridMenu* menu) {
    // Reset the current row size and position
    menu->current_row_size = 0;
    menu->current_row_position = menu->base.y;
    
    // Call the base menu cleanup to reposition all items
    CleanupMenu(&menu->base);
}

void UpdateGridMenu(GridMenu* menu, float delta_time) {
    // Update the base menu
    UpdateMenu(&menu->base, delta_time);
}

void DrawGridMenu(const GridMenu* menu) {
    // Draw the base menu
    DrawMenu(&menu->base);
} 